Mathlib/NumberTheory/Fermat.lean

```diff
@@ -0,0 +1,79 @@
+/-
+Copyright (c) 2024 Moritz Firsching. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Moritz Firsching
+-/
+import Mathlib.Algebra.BigOperators.Group.Finset
+import Mathlib.Algebra.Order.Ring.Basic
+import Mathlib.Algebra.Order.Star.Basic
+import Mathlib.Data.Nat.Prime.Defs
+import Mathlib.Tactic.Ring.RingNF
+
+/-!
+# Fermat numbers
+
+The Fermat numbers are a sequence of natural numbers defined as $2^{2^n} + 1$, for all natural
+numbers $n$.
+
+## Main theorems
+  `coprime_fermat_fermat`: two distinct fermat numbers are coprime.
+-/
+open Nat
+open scoped BigOperators
+
+/-- The Fermat numbers:
+the $n$-th Fermat number $F_n$ is defined as $2^{2^n} + 1$. -/
+def fermat : ℕ → ℕ := fun n ↦ 2 ^ (2 ^ n) + 1
+
+@[simp]
+theorem fermat_zero : fermat 0 = 3 := by rfl
+
+@[simp]
+theorem fermat_one : fermat 1 = 5 := by rfl
+
+@[simp]
+theorem fermat_two : fermat 2 = 17 := by rfl
+
+theorem strictMono_fermat : StrictMono fermat := by
+  apply strictMono_nat_of_lt_succ
+  simp only [fermat, add_lt_add_iff_right, Nat.pow_succ]
+  exact fun n => (pow_lt_pow_iff_right one_lt_two).mpr (by aesop)
+
+theorem two_lt_fermat {n : ℕ} : 2 < fermat n := by
+  cases n
+  · simp
+  · exact lt_of_succ_lt <| strictMono_fermat <| zero_lt_succ _
+
+theorem odd_fermat {n : ℕ} : Odd (fermat n) := by
+  rw [fermat, ← Nat.not_even_iff_odd, Nat.even_add_one, not_not, Nat.even_pow]
+  refine ⟨even_two, Nat.ne_of_gt (pow_pos zero_lt_two _)⟩
+
+@[simp]
+theorem fermat_product {n : ℕ} : ∏ k in Finset.range n, fermat k = fermat n - 2 := by
```

Comment:
a) Does this make sense as a `simp` lemma? I can't imagine you're going to come across the LHS much, even within this file.
b) I'd imagine `fermat_eq_prod_add_two` is the more useful lemma, since it avoids natural subtraction (and it's the one you use for `coprime_fermat_fermat`).

Reply:
a) not super sure, but it seems to be a somewhat simple lhs. For the next lemma in this file, we could use this simp. But I'm happy to remove it if you think
b) yeah, since I use `fermat_product` in the proof anyhow, I thought I might as well give it a name.

Reply:
removed `simp`